pkg/metrics/metrics.go

```diff
@@ -0,0 +1,147 @@
+	if condition {
+		m.prometheusCounter.Inc()
+	}
+}
+
+// Gauge is a wrapper of the prometheus gauge type
+type Gauge struct {
+	Name            string
+	Help            string
+	prometheusGauge prometheus.Gauge
+}
+
+// Collector returns a collector for the metric
+func (m Gauge) Collector() prometheus.Collector {
+	return m.prometheusGauge
+}
+
+// Set the gauge if it has been initialized
+func (m Gauge) Set(value float64) {
+	m.prometheusGauge.Set(value)
+}
+
+// counters
+var (
+	// UserSignupUniqueTotal should be incremented only the first time a user signup is created, there should be 1 for each unique user
+	UserSignupUniqueTotal = initCounter("user_signups_total", "Total number of unique User Signups")
+
+	// UserSignupProvisionedTotal should be incremented each time a user signup is provisioned, can be multiple times per user if they reactivate multiple times
+	UserSignupProvisionedTotal = initCounter("user_signups_provisioned_total", "Total number of Provisioned User Signups")
+
+	// UserSignupBannedTotal should be incremented each time a user signup is banned
+	UserSignupBannedTotal = initCounter("user_signups_banned_total", "Total number of Banned User Signups")
+
+	// UserSignupDeactivatedTotal should be incremented each time a user signup is deactivated, can be multiple times per user if they reactivate multiple times
+	UserSignupDeactivatedTotal = initCounter("user_signups_deactivated_total", "Total number of Deactivated User Signups")
+
+	// UserSignupAutoDeactivatedTotal should be incremented each time a user signup is automatically deactivated, can be multiple times per user if they reactivate multiple times
+	UserSignupAutoDeactivatedTotal = initCounter("user_signups_auto_deactivated_total", "Total number of Automatically Deactivated User Signups")
+)
+
+// gauges
+var (
+	// MasterUserRecordGauge should reflect the current number of master user records in the system
+	MasterUserRecordGauge = initGauge("master_user_record_current", "Current number of Master User Records")
+)
+
+// collections
+var (
+	allCounters = []*Counter{}
+	allGauges   = []*Gauge{}
+)
+
+func initCounter(name, help string) *Counter {
+	c := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: name,
+		Help: help,
+	})
+	m := &Counter{
+		Name:              name,
+		Help:              help,
+		prometheusCounter: c,
+	}
+	allCounters = append(allCounters, m)
+	return m
+}
+
+func initGauge(name, help string) *Gauge {
+	g := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: name,
+		Help: help,
+	})
+	m := &Gauge{
+		Name:            name,
+		Help:            help,
+		prometheusGauge: g,
+	}
+	allGauges = append(allGauges, m)
+	return m
+}
+
+// RegisterCustomMetrics registers the custom metrics
+func RegisterCustomMetrics() {
+	// register metrics
+	for _, m := range allCounters {
+		k8smetrics.Registry.MustRegister(m.prometheusCounter)
+	}
+	for _, m := range allGauges {
+		k8smetrics.Registry.MustRegister(m.prometheusGauge)
```

Comment:
Since you are using the `MustRegister`, could you please that it won't fail for rolling updates?  I remember that in some cases it failed with `AlreadyRegisteredError` but don't exactly remember what was the case :man_shrugging:

Reply:
I got that from Xavier's original PR: https://github.com/codeready-toolchain/host-operator/pull/286/files

@xcoulon maybe it's better for you to explain rather than me guessing?

Reply:
IIRC, I found this code in the Operator SDK, but I can't recall where exactly. 
Similar code exists in https://github.com/operator-framework/operator-sdk/blob/55f1b845abb7630064f79c3dbf79b66460caba21/internal/ansible/metrics/metrics.go#L53, but I don't know if in our case here we may have some issues with rolling updates.

Reply:
Let me do some investigation to see whether rolling updates would be a problem

Reply:
It looks like it's fine after a rolling update. The new pod is up and running and I can see that the counters are reset (as expected) and the `master_user_record_current` gauge is correct so I guess we're good.

According to https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#hdr-Advanced_Uses_of_the_Registry it should cause a panic if there's an error during registration so I would expect the pod to crashloop if the registration fails. I believe this is what we want because we shouldn't be registering metrics that are incompatible or inconsistent with already registered metrics.

Reply:
what do you mean by:

> the counters are reset 

what it looks like in the prometheus? Is there any drop, or it just increases the number, so we can easily make a time range across the update...?

Reply:
>the counters are reset

Sorry, I should have clarified that I meant when I check the metrics endpoint which Prometheus scrapes then the values for all the counters are 0 which is what I expected to see after the update.

>what it looks like in the prometheus? Is there any drop, or it just increases the number, so we can easily make a time range across the update...?

Prometheus handles counter resets so we can easily run prometheus queries like `increase(sandbox_user_signups_unique_total[1h])` which would tell us how many new signups there were within the last hour.

Reply:
Awesome, thanks for checking it and explaining it to me :wink:

Reply:
I still need to check our operator metrics in Prometheus. I couldn't get it to work with the Openshift built-in prometheus so I need to figure out the setup process for a separately installed Prometheus but I've used similar queries for other counters using the Openshift built-in prometheus so what I explained is based on that.